### DISCUSSION TITLE
How to map a union type in ManyToOne relation?

### DISCUSSION THREAD

**Discussion by fluffycondor** [doctrine/doctrine2#9510] [Support Questions] [answered]
Considering I have a property like this:

```php
class Message {
    #[ORM\ManyToOne]
    #[ORM\JoinColumn(name: 'receiver', referencedColumnName: 'id', nullable: false)]
    private User|ChatGroup $receiver;
}
```
If I try to make a migration for it, it fails with a `MappingException`:
> The association mapping 'receiver' misses the 'targetEntity' attribute.

I think about two ways to solve it: just store UuidInterface of the target entities, or implement some inheritance like UserMessage and GroupMessage to make this property's type just ReflectionNamedType like this:
```php
class UserMessage {
    private User $receiver;
}
class GroupMessage {
    private ChatGroup $receiver;
}
```
But both of these ways seems a little clunky to me.

Is there a way to map a union type now? If not, how to solve this problem in a proper way?

**Comment by derrabus** [comment, marked as answer]
Introduce a common superclass and configure a discriminator mapping there.

https://www.doctrine-project.org/projects/doctrine-orm/en/2.11/reference/inheritance-mapping.html

**Reply by fluffycondor** [reply]
Thank you for the answer!
Just to make sure I understand it correctly, should it looks something like this?
````php
#[Entity]
#[InheritanceType("SINGLE_TABLE")]
#[DiscriminatorColumn(name: "discr", type: "string")]
abstract class Message {}

class UserMessage extends Message {
    #[ManyToOne]
    #[JoinColumn(nullable: false)]
    private User $receiver;
}
class GroupMessage extends Message {
    #[ManyToOne]
    #[JoinColumn(nullable: false)]
    private ChatGroup $receiver;
}
````

**Reply by derrabus** [reply]
No, `User` and `ChatGroup` should have a common superclass.

**Reply by fluffycondor** [reply]
I carefully read the docs, but I have a problem with id. The documentation gives such example:
````php
/** @MappedSuperclass */
class Person {}

/** @Entity */
class Employee extends Person
{
    /** @Id @Column(type="integer") */
    private $id;
}
````

I tried to adopt it in my project, but I get a `MissingColumnException`:
`Column name "id" referenced for relation from App\Entity\ChatMessage towards App\Entity\MessageReceiver does not exist. `
It didn't exist indeed, so I did it like this:

````php
#[MappedSuperclass]
class MessageReceiver {
    #[Id, Column(type: 'uuid', unique: true)]
    protected UuidInterface $id;
}

#[Entity]
class User extends MessageReceiver
{
    #[Id, Column(type: 'uuid', unique: true)]
    protected UuidInterface $id;
}

#[Entity]
class ChatGroup extends MessageReceiver
{
    #[Id, Column(type: 'uuid', unique: true)]
    protected UuidInterface $id;
}

#[Entity]
class ChatMessage
{
    #[ManyToOne(targetEntity: MessageReceiver::class)]
    #[JoinColumn(name: 'receiver', nullable: false)]
    public readonly User|ChatGroup $receiver;
}
````
But I still get the same exception. I've executed `cache:clear-metadata` command, but no luck.

Looks like a have to write a DiscriminatiorMap somewhere? Could you please give me an example how to make a ManyToOne relation to a MappedSuperclass?

**Reply by derrabus** [reply]
Does that help?

```php
#[Entity]
#[InheritanceType("JOINED")]
#[DiscriminatorColumn('type')]
#[DiscriminatorMap([
    'user' => User::class,
    'group' => ChatGroup::class,
])]
abstract class MessageReceiver
{
    #[Id, Column, GeneratedValue]
    protected ?int $id = null;
}

#[Entity]
class User extends MessageReceiver
{
}

#[Entity]
class ChatGroup extends MessageReceiver
{
}

#[Entity]
class ChatMessage
{
    #[Column, Id, GeneratedValue]
    public ?int $id = null;

    #[ManyToOne, JoinColumn(nullable: false)]
    public MessageReceiver $receiver;
}
```

**Reply by fluffycondor** [reply]
It works like a charm, thank you very much!

One last question: is there a way to avoid such inheritance, perhaps using interfaces? It seems a little weird to me that `User` extends `MessageReceiver` to make it work :sweat_smile: What if I need to map another ManyToOne `User|Something` union type somewhere else, or `User` would already be inherited from some parent class?

**Reply by derrabus** [reply]
The problem is that the ORM needs an efficient way to decide which class to load when hydrating a record from the database. This is what the discriminator column is for. The way Doctrine's metadata system works, metadata has to be attached to an actual class. So right now, the common superclass is the only way to configure this.
